File: application/backend-credit-card-enrollment/backend-typescript/src/common/util/MongoInitializer.ts
import { injectable, inject } from 'tsyringe';
import { MongoClient } from 'mongodb';
import { MongoSessionPool } from './MongoSessionPool';
import { log } from './Logger';

@injectable()
export class MongoInitializer {
    private readonly client: MongoClient;

    constructor(
        @inject(MongoSessionPool) private readonly sessionPool: MongoSessionPool,
        @inject("mongoDatabaseName") private readonly databaseName: string
    ) {
        this.client = this.sessionPool.getClient();
    }

    async initialize(): Promise<void> {
        log.info('Initializing MongoDB collections and indexes...');

        try {
            await this.client.connect();
            const db = this.client.db(this.databaseName);

            // Create collections
            log.info('Creating collections...');
            await Promise.all([
                this.ensureCollection(db, 'CreditCard_Enrollment_Enrollment'),
                this.ensureCollection(db, 'CreditCard_Enrollment_ProductName'),
                this.ensureCollection(db, 'CreditCard_Enrollment_ProductActiveStatus')
            ]);
            log.info('Collections created successfully');

            // Create indexes
            log.info('Creating indexes...');
            await this.createIndexes(db);
            log.info('Indexes created successfully');

        } catch (error) {
            log.error('Error initializing MongoDB:', error as Error);
            throw error;
        }
    }

    private async ensureCollection(db: any, collectionName: string): Promise<void> {
        try {
            const collections = await db.listCollections({ name: collectionName }).toArray();
            if (collections.length === 0) {
                await db.createCollection(collectionName);
                log.debug(`Collection ${collectionName} created`);
            } else {
                log.debug(`Collection ${collectionName} already exists`);
            }
        } catch (error) {
            log.error(`Error ensuring collection ${collectionName}:`, error as Error);
            throw error;
        }
    }

    private async createIndexes(db: any): Promise<void> {
        try {
            const enrollmentCollection = db.collection('CreditCard_Enrollment_Enrollment');

            await enrollmentCollection.createIndex(
                { userId: 1 },
                {
                    background: true,
                    name: 'userId_asc'
                }
            );
            log.debug('Index created on CreditCard_Enrollment_Enrollment.userId');
        } catch (error) {
            log.error('Error creating indexes:', error as Error);
            throw error;
        }
    }
}

File: application/backend-credit-card-enrollment/backend-c#/Common/Util/MongoInitializer.cs
using MongoDB.Driver;

namespace CreditCardEnrollment.Common.Util;

public class MongoInitializer
{
    private readonly MongoSessionPool _sessionPool;
    private readonly string _databaseName;
    private readonly ILogger<MongoInitializer> _logger;

    public MongoInitializer(
        MongoSessionPool sessionPool,
        string databaseName,
        ILogger<MongoInitializer> logger)
    {
        _sessionPool = sessionPool;
        _databaseName = databaseName;
        _logger = logger;
    }

    public void Initialize()
    {
        using var session = _sessionPool.StartSession();
        var database = session.Client.GetDatabase(_databaseName);

        try
        {
            _logger.LogInformation("Creating collections");
            CreateCollectionIfNotExists(database, "CreditCard_Enrollment_Enrollment");
            CreateCollectionIfNotExists(database, "CreditCard_Enrollment_ProductName");
            CreateCollectionIfNotExists(database, "CreditCard_Enrollment_ProductActiveStatus");
            _logger.LogInformation("Created collections");

            _logger.LogInformation("Creating indexes");
            var enrollmentCollection = database.GetCollection<object>("CreditCard_Enrollment_Enrollment");
            var indexKeysDefinition = Builders<object>.IndexKeys.Ascending("userId");
            enrollmentCollection.Indexes.CreateOne(new CreateIndexModel<object>(indexKeysDefinition));
            _logger.LogInformation("Created indexes");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error initializing MongoDB");
            throw;
        }
    }

    private void CreateCollectionIfNotExists(IMongoDatabase database, string collectionName)
    {
        try
        {
            var collections = database.ListCollectionNames().ToList();
            if (!collections.Contains(collectionName))
            {
                database.CreateCollection(collectionName);
                _logger.LogInformation("Created collection {CollectionName}", collectionName);
            }
            else
            {
                _logger.LogInformation("Collection {CollectionName} already exists", collectionName);
            }
        }
        catch (Exception e)
        {
            _logger.LogWarning(e, "Error creating collection {CollectionName}", collectionName);
            throw;
        }
    }
}